csharp/ql/src/experimental/Security Features/CWE-348/ClientSuppliedIpUsedInSecurityCheckLib.qll

```diff
@@ -0,0 +1,170 @@
+private import semmle.code.csharp.security.dataflow.flowsources.Remote
+private import semmle.code.csharp.frameworks.microsoft.AspNetCore
+private import semmle.code.csharp.frameworks.System
+private import semmle.code.csharp.security.dataflow.SqlInjectionQuery
+
+/** A data flow source for ip address forgery vulnerabilities. */
+abstract class ClientSuppliedIpUsedInSecurityCheckSource extends DataFlow::Node { }
+
+/** A data flow sink for ip address forgery vulnerabilities. */
+abstract class ClientSuppliedIpUsedInSecurityCheckSink extends DataFlow::ExprNode { }
+
+/** A data flow sanitizer for ip address forgery vulnerabilities. */
+abstract class ClientSuppliedIpUsedInSecurityCheckSanitizer extends DataFlow::ExprNode { }
+
+/**
+ * A data flow source of the client ip obtained according to the remote endpoint identifier specified
+ * (`X-Forwarded-For`, `X-Real-IP`, `Proxy-Client-IP`, etc.) in the header.
+ *
+ * For example: `Request.Headers["X-Forwarded-For"]` or `request.ServerVariables["HTTP_X_FORWARDED_FOR"]`.
+ */
+private class AspNetQueryStringSource extends ClientSuppliedIpUsedInSecurityCheckSource,
+  DataFlow::ExprNode {
+  AspNetQueryStringSource() {
+    exists(RemoteFlowSource rfs, RefType rt, IndexerAccess ia, Call c |
+      rfs.asExpr() = ia.getQualifier() and
+      ia.getTarget() = rt.getAnIndexer() and
+      c.getTarget() = rt.getAnIndexer().getGetter() and
+      c.getArgument(0).(StringLiteral).getValue().toLowerCase() = clientIpParameterName() and
+      this.getExpr() = c
+    )
+    or
+    //  `Request.Headers.TryGetValue("X-Forwarded-For", out var xHeader)`
+    exists(RemoteFlowSource rfs, MethodCall mc |
+      rfs.asExpr() = mc.getQualifier() and
+      mc.getArgument(0).(StringLiteral).getValue().toLowerCase() = clientIpParameterName() and
+      this.getExpr() = mc
```

Comment:
So shouldn't it be `xHeader` that is the source in this case, and not the method call itself?